File: communityInterview/download_transcripts.py
#!/usr/bin/env python3
"""
Download YouTube transcripts from a list of video URLs.
Requires: pip install youtube-transcript-api
"""

import re
import os
from pathlib import Path
from youtube_transcript_api import YouTubeTranscriptApi
from youtube_transcript_api._errors import TranscriptsDisabled, NoTranscriptFound


def extract_video_id(url):
    """Extract video ID from YouTube URL."""
    patterns = [
        r'(?:v=|\/)([0-9A-Za-z_-]{11}).*',
        r'(?:embed\/)([0-9A-Za-z_-]{11})',
        r'^([0-9A-Za-z_-]{11})$'
    ]

    for pattern in patterns:
        match = re.search(pattern, url)
        if match:
            return match.group(1)
    return None


def download_transcript(video_id, output_dir):
    """Download transcript for a single video."""
    try:
        # Get transcript (tries to get English first, then any available)
        transcript_list = YouTubeTranscriptApi.list_transcripts(video_id)

        # Try to find English transcript first
        try:
            transcript = transcript_list.find_transcript(['en'])
        except:
            # If no English, get the first available
            transcript = transcript_list.find_generated_transcript(['en'])

        # Fetch the actual transcript
        transcript_data = transcript.fetch()

        # Format transcript as readable text
        full_text = "\n".join([entry['text'] for entry in transcript_data])

        # Save to file
        output_file = output_dir / f"{video_id}.txt"
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(f"Video ID: {video_id}\n")
            f.write(f"URL: https://www.youtube.com/watch?v={video_id}\n")
            f.write("-" * 80 + "\n\n")
            f.write(full_text)

        print(f"✓ Downloaded: {video_id}")
        return True

    except TranscriptsDisabled:
        print(f"✗ Transcripts disabled: {video_id}")
        return False
    except NoTranscriptFound:
        print(f"✗ No transcript found: {video_id}")
        return False
    except Exception as e:
        print(f"✗ Error for {video_id}: {str(e)}")
        return False


def main():
    # Setup paths
    script_dir = Path(__file__).parent
    input_file = script_dir / "The Wellbeing Youtube Video - Voices of Wellbeing.txt"
    output_dir = script_dir / "transcripts"

    # Create output directory
    output_dir.mkdir(exist_ok=True)

    # Read video URLs
    print(f"Reading video list from: {input_file}")
    with open(input_file, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    # Extract video IDs
    video_ids = []
    for line in lines:
        # Extract URL (skip the numbering)
        url_match = re.search(r'https?://[^\s]+', line)
        if url_match:
            url = url_match.group(0)
            video_id = extract_video_id(url)
            if video_id:
                video_ids.append(video_id)

    # Remove duplicates while preserving order
    video_ids = list(dict.fromkeys(video_ids))

    print(f"\nFound {len(video_ids)} unique videos")
    print(f"Output directory: {output_dir}\n")

    # Download transcripts
    success_count = 0
    for i, video_id in enumerate(video_ids, 1):
        print(f"[{i}/{len(video_ids)}] Processing {video_id}...", end=" ")
        if download_transcript(video_id, output_dir):
            success_count += 1

    # Summary
    print(f"\n{'='*80}")
    print(f"Complete! Successfully downloaded {success_count}/{len(video_ids)} transcripts")
    print(f"Transcripts saved to: {output_dir}")


if __name__ == "__main__":
    main()

File: communityInterview/transcripts/download_transcripts.js
#!/usr/bin/env node
/**
 * WBP Community Interview Transcript Downloader
 * Uses the same youtube-transcript library as transcript-explorer
 * Standalone version - no database, just saves to files
 */

const fs = require('fs');
const path = require('path');

// Check if youtube-transcript is installed
let YoutubeTranscript;
try {
  YoutubeTranscript = require('youtube-transcript').YoutubeTranscript;
} catch (e) {
  console.error('youtube-transcript not installed. Installing now...');
  require('child_process').execSync('npm install youtube-transcript', { stdio: 'inherit' });
  YoutubeTranscript = require('youtube-transcript').YoutubeTranscript;
}

function extractVideoId(urlOrId) {
  // Same logic as transcript-explorer
  const m = urlOrId.match(/(?:v=|youtu\.be\/|\/shorts\/)([A-Za-z0-9_-]{11})/);
  if (m) return m[1];
  if (/^[A-Za-z0-9_-]{11}$/.test(urlOrId)) return urlOrId;
  return urlOrId;
}

async function downloadTranscript(url, outputDir) {
  const videoId = extractVideoId(url);
  const outputFile = path.join(outputDir, `${videoId}_transcript.txt`);

  // Check if already downloaded
  if (fs.existsSync(outputFile)) {
    console.log(`  ⏭  Already downloaded`);
    return true;
  }

  // Try a few sensible fallbacks (same as transcript-explorer)
  let items = [];
  const tries = [
    async () => YoutubeTranscript.fetchTranscript(videoId),
    async () => YoutubeTranscript.fetchTranscript(videoId, { lang: 'en' }),
    async () => YoutubeTranscript.fetchTranscript(videoId, { lang: 'en-US' }),
  ];

  for (const t of tries) {
    try {
      items = await t();
      if (items?.length) break;
    } catch (err) {
      // Try next fallback
    }
  }

  if (!items?.length) {
    console.log(`  ✗ No transcript available`);
    return false;
  }

  // Format transcript
  const lines = [
    `Video ID: ${videoId}`,
    `URL: ${url}`,
    `Fetched: ${new Date().toISOString()}`,
    `Segments: ${items.length}`,
    '='.repeat(80),
    '',
    'TRANSCRIPT:',
    ''
  ];

  // Add timestamped text
  for (const item of items) {
    const start = item.offset || 0;
    const minutes = Math.floor(start / 60);
    const seconds = Math.floor(start % 60);
    const timestamp = `${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
    lines.push(`[${timestamp}] ${item.text}`);
  }

  // Save to file
  fs.writeFileSync(outputFile, lines.join('\n'), 'utf-8');
  console.log(`  ✅ Saved: ${items.length} segments`);
  return true;
}

async function main() {
  const inputFile = path.join(__dirname, '..', 'The Wellbeing Youtube Video - Voices of Wellbeing.txt');
  const outputDir = __dirname;

  if (!fs.existsSync(inputFile)) {
    console.error(`Input file not found: ${inputFile}`);
    process.exit(1);
  }

  console.log(`
╔══════════════════════════════════════════════════════════╗
║  WBP Community Interview Transcript Downloader          ║
║  Using youtube-transcript (from transcript-explorer)    ║
╚══════════════════════════════════════════════════════════╝

Input: ${path.basename(inputFile)}
Output: ${outputDir}/
  `);

  // Read URLs
  const content = fs.readFileSync(inputFile, 'utf-8');
  const lines = content.split('\n');

  // Extract URLs
  const urls = [];
  for (const line of lines) {
    const match = line.match(/https?:\/\/[^\s]+/);
    if (match) {
      urls.push(match[0]);
    }
  }

  // Remove duplicates
  const uniqueUrls = [...new Set(urls)];

  console.log(`Found ${uniqueUrls.length} unique videos\n`);
  console.log('='.repeat(60));

  // Download transcripts
  let successCount = 0;
  for (let i = 0; i < uniqueUrls.length; i++) {
    const url = uniqueUrls[i];
    const videoId = extractVideoId(url);

    console.log(`\n[${i + 1}/${uniqueUrls.length}] ${videoId}`);
    console.log(`  ${url}`);

    try {
      if (await downloadTranscript(url, outputDir)) {
        successCount++;
      }
    } catch (err) {
      console.log(`  ✗ Error: ${err.message}`);
    }

    // Small delay to be nice
    if (i < uniqueUrls.length - 1) {
      await new Promise(resolve => setTimeout(resolve, 1000));
    }
  }

  // Summary
  console.log(`\n${'='.repeat(60)}`);
  console.log(`Complete! Downloaded ${successCount}/${uniqueUrls.length} transcripts`);
  console.log(`Transcripts saved to: ${outputDir}/`);
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});
